File: src/main/java/site/overwrite/auditranscribe/spectrogram/vqt_from_scratch/vqt.py
# IMPORTS
import numpy as np  # Ideally we shouldn't rely on many methods from Numpy as they aren't available in Java
import librosa

from vqt_from_scratch.audio import *
from vqt_from_scratch.helpers import *
from vqt_from_scratch.utils import *
from vqt_from_scratch.windowing import *

# CONSTANTS
# Resampling bandwidths as percentage of Nyquist Frequency
# (Taken from https://github.com/bmcfee/resampy/blob/0.2.1/resampy/filters.py)
BW_BEST = 0.9475937167399596
BW_FASTEST = 0.85


# CLASSES
class VQT:  # In java we can implement this as a static class
    """
    VQT Class that contains VQT methods
    
    Adapted largely from https://librosa.org/doc/main/_modules/librosa/core/constantq.html#vqt.

    Reference: Schoerkhuber, Christian, and Anssi Klapuri.
        "Constant-Q transform toolbox for music processing."
        7th Sound and Music Computing Conference, Barcelona, Spain. 2010.
    """

    def __init__(self):  # Don't need this in Java
        pass

    # Public methods
    def vqt(self, y, sr, hop_length=512, fmin=None, n_bins=168, bins_per_octave=24, gamma=None, window="hann"):  # Can be made static in Java
        # Compute number of octaves that we are processing and the number of filters
        n_octaves = self.get_num_octaves(n_bins, bins_per_octave)
        n_filters = min(bins_per_octave, n_bins)  # What is a 'filter'??? Todo: wouldn't this always be `bins_per_octave`?
        
        # Assign values to `None` kwargs
        if fmin is None:
            # Minimum frequency is C1 by default
            fmin = librosa.note_to_hz("C1")  # This function should be implemented under Misc utils; tuned to A440

        #tuning = 0  # Although librosa supports tuning estimation, we shouldn't need to estimate this
        dtype = librosa.util.dtype_r2c(y.dtype)  # Converts the 'real' datatype of `y` to a 'complex' datatype

        # Get the VQT frequency bins
        freqs = self.get_freq_bins(n_bins, bins_per_octave, fmin)

        # Get the frequencies for the top octave
        freqs_top = [0] * bins_per_octave  # In Java do smth like `double[] freqs_top = new double[bins_per_octave]`

        for bin_index in range(bins_per_octave):
            freqs_top[bin_index] = freqs[n_bins - bins_per_octave + bin_index]

        # Get highest frequency
        fmax_t = freqs_top[bins_per_octave - 1]  # Why on earth did they name this variable `fmax_t`?????     

        # Calculate the relative difference in frequency between any two consecutive bands, alpha
        alpha = compute_alpha(bins_per_octave)

        # Compute wavelet lengths
        lengths, filter_cutoff = compute_wavelet_lengths(freqs, sr, window, gamma, alpha)

        # Determine required resampling quality
        nyquist_freq = sr / 2.0  # See Nyquist–Shannon sampling theorem

        if filter_cutoff > nyquist_freq:
            raise ParameterError(
                f"Wavelet basis with max frequency {fmax_t} would exceed the Nyquist frequency {nyquist_freq}. "
                "Try reducing the number of frequency bins."
            )

        # Resample audio
        if filter_cutoff < BW_FASTEST * nyquist_freq:
            resample_type = "kaiser_fast"
        else:
            resample_type = "kaiser_best"
    
        y, sr, hop_length = self.early_downsample(y, sr, hop_length, resample_type, n_octaves, nyquist_freq, filter_cutoff)

        # Define VQT response array
        vqt_resp = []  # Has length `n_octaves`; todo: find the 'internal' shape

        # Skip this block for now
        # (Todo: update this comment; this really isn't that descriptive)
        oct_start = 0
        
        if resample_type != "kaiser_fast":
            # Get the top frequencies
            # freqs_top = freqs[-n_filters:]

            freqs_top = [0] * n_filters  # In Java do smth like `double[] freqs_top = new double[n_filters]`

            for bin_index in range(n_filters):
                freqs_top[bin_index] = freqs[n_bins - n_filters + bin_index]

            # Do the top octave before resampling to allow for fast resampling
            fft_basis, n_fft, _ = self.vqt_filter_fft(
                sr,
                freqs_top,
                1,  # Filter scale is 1
                1,  # Norm is 1
                window=window,
                gamma=gamma,
                dtype=dtype,
                alpha=alpha,
            )

            # Compute the VQT filter response and append it to the stack
            vqt_resp.append(
                self.vqt_response(y, n_fft, hop_length, fft_basis, dtype=dtype)
            )

            oct_start = 1

            resample_type = "kaiser_fast"

        # Iterate down the octaves
        my_y, my_sr, my_hop = y, sr, hop_length  # Todo: rename variables

        for i in range(oct_start, n_octaves):  # This starts from the HIGHEST frequencies and goes down
            # Get the frequencies of the current octave
            # Fixme: This may be incorrect with early downsampling
            freqs_oct = [0] * n_filters  # In Java do smth like `double[] freqs_oct = new double[n_filters]`

            for bin_index in range(n_filters):
                freqs_oct[bin_index] = freqs[n_bins - n_filters * (i + 1) + bin_index]

            # Get the FFT basis and the `n_fft` for this octave
            fft_basis, n_fft, _ = self.vqt_filter_fft(
                my_sr,
                freqs_oct,
                1,  # Filter scale is 1
                1,  # Norm is 1
                window=window,
                gamma=gamma,
                dtype=dtype,
                alpha=alpha,
            )

            # Re-scale the filters to compensate for downsampling
            fft_basis[:] *= np.sqrt(sr / my_sr)

            # Compute the VQT filter response and append to the stack
            vqtresponse = self.vqt_response(my_y, n_fft, my_hop, fft_basis, dtype=dtype)  # Possible error here

            vqt_resp.append(
                vqtresponse
            )

            #  Update variables
            if my_hop % 2 == 0:
                my_hop //= 2
                my_sr /= 2.0
                my_y = resample(
                    my_y, sr_orig=2, sr_new=1, res_type="kaiser_fast", scale=True
                )

        V = self.trim_stack(vqt_resp, n_bins, dtype)

        # Recompute lengths here because early downsampling may have changed our sampling rate
        lengths, _ = compute_wavelet_lengths(
            freqs=freqs,
            sr=sr,
            window=window,
            filter_scale=1,
            gamma_value=gamma,
            fallback_alpha=alpha,
        )

        # Scale `V` back to normal
        for i in range(V.shape[0]):
            scale_factor = np.sqrt(lengths[i])

            for j in range(V.shape[1]):
                V[i][j] /= scale_factor

        # Return VQT matrix
        return V


    # Private methods
    @staticmethod
    def get_num_octaves(n_bins, bins_per_octave):
        return int(np.ceil(float(n_bins) / bins_per_octave))  # `np.ceil` in java is `Math.ceil`


    @staticmethod
    def get_freq_bins(n_bins, bins_per_octave, fmin):
        frequencies = [0] * n_bins  # In Java do smth like `double[] frequencies = new double[n_bins]`

        for bin_index in range(n_bins):
            # Calculate the frequency of the current frequency bin
            freq = fmin * (2.0 ** (bin_index / bins_per_octave))

            # Append it to the list of frequencies
            frequencies[bin_index] = freq

        return np.array(frequencies)  # Converted to Numpy array for python consistency

    @staticmethod
    def early_downsample(y, sr, hop_length, res_type, n_octaves, nyquist, filter_cutoff):
        # Compute the number of early downsampling operations
        # (Note: all numpy operations here can be replaced by `Math` class)
        downsample_count1 = max(0, int(np.ceil(np.log2(BW_FASTEST * nyquist / filter_cutoff)) - 1) - 1)
        downsample_count2 = max(0, num_two_factors(hop_length) - n_octaves + 1)

        downsample_count = min(downsample_count1, downsample_count2)

        # Actually perform the downsampling
        if downsample_count > 0 and res_type == "kaiser_fast":
            # Compute how much to downsample by
            downsample_factor = 2 ** downsample_count

            # Check if the signal can actually be downsampled
            if y.shape[-1] < downsample_factor:
                raise ParameterError(f"Input signal length={len(y):d} is too short for {n_octaves:d}-octave CQT")

            # Downsample `hop_length` and the sample rate
            hop_length //= downsample_factor
            new_sr = sr / float(downsample_factor)

            # Downsample audio sample
            y = resample(y, sr_orig=sr, sr_new=new_sr, res_type=res_type, scale=True)
            sr = new_sr

        return y, sr, hop_length

    @staticmethod
    def vqt_filter_fft(sr, freqs, filter_scale, norm, window="hann", gamma=0.0, dtype=np.complex64, alpha=None):  # Todo: change name; this is a bad name
        """
        Generate the frequency domain variable-Q filter basis.

        Todo: update description
        """

        # Get the frequency and lengths of the wavelet basis
        basis, lengths = compute_wavelet_basis(
            freqs=freqs,
            sr=sr,
            filter_scale=filter_scale,
            norm=norm,
            pad_fft=True,
            window=window,
            gamma=gamma,
            alpha=alpha,
        )

        # Number of FFT bins is the second element of the shape of the basis
        n_fft = basis.shape[1]  # Todo: find a way to efficiently extract this in Java

        # Re-normalize bases with respect to the FFT window length
        for i in range(basis.shape[0]):
            normalisation_factor = lengths[i] / float(n_fft)

            for j in range(basis.shape[1]):
                basis[i][j] = basis[i][j] * normalisation_factor

        # FFT and retain only the non-negative frequencies
        fft_basis = np.fft.fft(basis, n=n_fft, axis=1)
        fft_basis = fft_basis[:, : (n_fft // 2) + 1]

        # Return required data
        return fft_basis, n_fft, lengths

    @staticmethod
    def vqt_response(y, n_fft, hop_length, fft_basis, dtype=None):
        """
        Compute the filter response with a target STFT hop.

        Assume `y` is single channeled.
        """

        # Compute the STFT matrix
        """
        Note:
        By default, `mode` is "constant" which means that the signal `y` is padded on BOTH SIDES with zeros until its
        length is a power of 2. The Java implementation of STFT should also use this.

        Todo:
        - Implement "ones" window
        """

        D = librosa.stft(
            y, n_fft=n_fft, hop_length=hop_length, window="ones", pad_mode="constant", dtype=dtype
        )

        # Define the output matrix
        output_flat = fft_basis.dot(D)  # This is matrix multiplication; todo: support it

        # Output
        return output_flat

    @staticmethod
    def trim_stack(cqt_resp, n_bins, dtype):  # Todo: rename `cqt_resp`
        """
        Helper function to trim and stack a collection of VQT responses.

        Assumes that `cqt_resp` is a numpy array with a shape like
            (X, Y, Z)
        This function returns a matrix with shape
            (X * Y, Z)
        where X * Y = n_bins
        """

        # Get maximum column
        # (Todo: find out why they use `min` instead of `max`)
        max_col = min(c_i.shape[-1] for c_i in cqt_resp)

        # Generate the output shape of the VQT matrix
        shape = (n_bins, max_col)

        # Generate output VQT matrix
        cqt_out = np.empty(shape, dtype=dtype, order="F")  # Todo: rename the variable

        # Copy per-octave data into output array
        end = n_bins

        for c_i in cqt_resp:  # Note: the FIRST element is that of the HIGHEST FREQUENCY bins
            # By default, take the whole octave
            n_oct = c_i.shape[-2]
            
            # If the whole octave is more than we can fit, take the highest bins from `c_i`
            if end < n_oct:  # Todo: check if this case ever occurs
                cqt_out[:end, :] = c_i[-end:, :max_col]
            else:
                cqt_out[end - n_oct : end, :] = c_i[:, :max_col]

            end -= n_oct

        return cqt_out

File: src/main/java/site/overwrite/auditranscribe/spectrogram/spectral_representations/VQT.java
/*
 * VQT.java
 * Description: Class that implements the Variable Q-Transform (VQT) algorithm.
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public Licence as published by the Free Software Foundation, either version 3 of the
 * Licence, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public Licence for more details.
 *
 * You should have received a copy of the GNU General Public Licence along with this program. If
 * not, see <https://www.gnu.org/licenses/>
 *
 * Copyright © AudiTranscribe Team
 */

package site.overwrite.auditranscribe.spectrogram.spectral_representations;

import site.overwrite.auditranscribe.audio.Audio;
import site.overwrite.auditranscribe.audio.Filter;
import site.overwrite.auditranscribe.audio.WindowFunction;
import site.overwrite.auditranscribe.generic.ClassWithLogging;
import site.overwrite.auditranscribe.generic.exceptions.ValueException;
import site.overwrite.auditranscribe.generic.tuples.Pair;
import site.overwrite.auditranscribe.generic.tuples.Triple;
import site.overwrite.auditranscribe.misc.Complex;
import site.overwrite.auditranscribe.misc.CustomTask;
import site.overwrite.auditranscribe.spectrogram.SpectralHelpers;
import site.overwrite.auditranscribe.spectrogram.Wavelet;
import site.overwrite.auditranscribe.utils.ArrayUtils;
import site.overwrite.auditranscribe.utils.MathUtils;

import java.security.InvalidParameterException;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Level;

/**
 * Class that implements the Variable Q-Transform (VQT) algorithm.
 *
 * @implNote Adapted largely from
 * <a href="https://librosa.org/doc/main/_modules/librosa/core/constantq.html#vqt">Librosa's
 * Implementation</a> of the VQT.
 * @see <a href="https://core.ac.uk/download/pdf/144846462.pdf">This paper</a> by Schoerkhuber,
 * Christian, and Anssi Klapuri. "Constant-Q transform toolbox for music processing." 7th Sound and
 * Music Computing Conference, Barcelona, Spain. 2010.
 */
public final class VQT extends ClassWithLogging {
    // Constants
    static final double BW_FASTEST = 0.85;

    private VQT() {
        // Private constructor to signal this is a utility class
    }

    // Public methods

    /**
     * Variable-Q Transform function.
     *
     * @param y              Audio time series.
     * @param sr             Sample rate of the audio.
     * @param hopLength      Number of samples between successive VQT columns.
     * @param fmin           Minimum frequency.
     * @param numBins        Number of frequency bins, starting at <code>fmin</code>.
     * @param binsPerOctave  Number of bins per octave.
     * @param tuning         Tuning offset in fractions of a bin.<br>
     *                       If <code>tuning = Double.NaN</code>, the <code>tuning</code> value will
     *                       be automatically estimated from the signal.<br>
     *                       The minimum frequency of the resulting Q-transform will be modified to
     *                       <code>fmin * Math.pow(2., tuning / binsPerOctave)</code>.
     * @param gamma          Bandwidth offset for determining filter lengths. <code>gamma = 0</code>
     *                       means that the gamma value will be derived automatically.
     * @param windowFunction Window function to apply to the basis filters.
     * @param task           The <code>CustomTask</code> object that is handling the generation.
     *                       Pass in <code>null</code> if no such task is being used.
     * @return Variable-Q value each frequency at each time.
     * @throws ValueException If: <ul>
     *                        <li>
     *                        The number of frequency bins is negative or zero.
     *                        </li>
     *                        <li>
     *                        The number of bins per octave is negative or zero.
     *                        </li>
     *                        <li>
     *                        The number of bins is not a multiple of the number of bins per octave.
     *                        </li>
     *                        <li>
     *                        The current number of frequency bins results in the highest frequency
     *                        exceeding the Nyquist frequency.
     *                        </li>
     *                        </ul>
     */
    public static Complex[][] vqt(
            double[] y, double sr, int hopLength, double fmin, int numBins, int binsPerOctave, double tuning,
            double gamma, WindowFunction windowFunction, CustomTask<?> task
    ) {
        return vqt(y, sr, hopLength, fmin, numBins, binsPerOctave, tuning, gamma, windowFunction, task, false);
    }

    /**
     * Variable-Q Transform function.
     *
     * @param y              Audio time series.
     * @param sr             Sample rate of the audio.
     * @param hopLength      Number of samples between successive Q-Transform columns.
     * @param fmin           Minimum frequency.
     * @param numBins        Number of frequency bins, starting at <code>fmin</code>.
     * @param binsPerOctave  Number of bins per octave.
     * @param tuning         Tuning offset in fractions of a bin.<br>
     *                       If <code>tuning = Double.NaN</code>, the <code>tuning</code> value will
     *                       be automatically estimated from the signal.<br>
     *                       The minimum frequency of the resulting Q-transform will be modified to
     *                       <code>fmin * Math.pow(2., tuning / binsPerOctave)</code>.
     * @param gamma          Bandwidth offset for determining filter lengths. If <code>isCQT</code>is
     *                       true and <code>gamma = 0</code>, produces the Constant-Q Transform
     *                       (CQT). Otherwise, <code>gamma = 0</code> means that the gamma value will
     *                       be derived automatically.
     * @param windowFunction Window function to apply to the basis filters.
     * @param task           The <code>CustomTask</code> object that is handling the generation.
     *                       Pass in <code>null</code> if no such task is being used.
     * @param isCQT          Whether this is a Constant-Q Transform or not.
     * @return Variable-Q value each frequency at each time.
     * @throws ValueException If: <ul>
     *                        <li>
     *                        The number of frequency bins is negative or zero.
     *                        </li>
     *                        <li>
     *                        The number of bins per octave is negative or zero.
     *                        </li>
     *                        <li>
     *                        The number of bins is not a multiple of the number of bins per octave.
     *                        </li>
     *                        <li>
     *                        The current number of frequency bins results in the highest frequency
     *                        exceeding the Nyquist frequency.
     *                        </li>
     *                        </ul>
     */
    public static Complex[][] vqt(
            double[] y, double sr, int hopLength, double fmin, int numBins, int binsPerOctave, double tuning,
            double gamma, WindowFunction windowFunction, CustomTask<?> task, boolean isCQT
    ) {
        // Validate parameters
        if (numBins <= 0) {
            throw new ValueException("Number of frequency bins cannot be negative or zero.");
        }

        if (binsPerOctave <= 0) {
            throw new ValueException("Number of bins per octave cannot be negative or zero");
        }

        if (numBins % binsPerOctave != 0) {
            throw new ValueException("Number of bins is not a multiple of the number of bins per octave.");
        }

        // Estimate tuning value from signal if needed
        if (Double.isNaN(tuning)) {
            tuning = SpectralHelpers.estimateTuning(y, sr);
        }

        // Apply tuning correction
        fmin = fmin * Math.pow(2., tuning / binsPerOctave);

        // Compute number of octaves that we are processing
        int numOctaves = getNumOctaves(numBins, binsPerOctave);

        // Get the VQT frequency bins
        double[] freqs = FrequencyBins.getQTFreqBins(numBins, binsPerOctave, fmin);

        // Get the frequencies for the top octave
        double[] freqsTop = new double[binsPerOctave];
        System.arraycopy(freqs, numBins - binsPerOctave, freqsTop, 0, binsPerOctave);

        // Calculate the relative difference in frequency between any two consecutive bands, alpha
        double alpha = SpectralHelpers.computeAlpha(binsPerOctave);

        // Compute filter cutoff
        Pair<double[], Double> waveletLengthsResponse = Wavelet.computeWaveletLengths(
                freqs, sr, windowFunction, 1., isCQT, gamma, alpha
        );
        double filterCutoff = waveletLengthsResponse.value1();

        // Determine required resampling quality
        double nyquistFrequency = sr / 2;

        if (filterCutoff > nyquistFrequency) {
            throw new InvalidParameterException(
                    "Wavelet basis with max frequency " + freqsTop[binsPerOctave - 1] +
                            " would exceed the Nyquist frequency " + nyquistFrequency +
                            ". Try reducing the number of frequency bins."
            );
        }

        // Resample audio
        Filter filter;
        if (filterCutoff < BW_FASTEST * nyquistFrequency) {
            filter = Filter.KAISER_FAST;
        } else {
            filter = Filter.KAISER_BEST;
        }

        Triple<double[], Double, Integer> earlyDownsampleResponse = earlyDownsample(
                y, sr, hopLength, filter, numOctaves, nyquistFrequency, filterCutoff
        );
        y = earlyDownsampleResponse.value0();
        sr = earlyDownsampleResponse.value1();
        hopLength = earlyDownsampleResponse.value2();

        // Define VQT response array
        List<Complex[][]> vqtResponses = new ArrayList<>();

        // Handle first octave specially if the filter type is NOT `KAISER_FAST`
        int startingOctave = 0;

        if (filter != Filter.KAISER_FAST) {
            // Get the frequencies of the top octave
            double[] freqsOct = new double[binsPerOctave];
            System.arraycopy(freqs, numBins - binsPerOctave, freqsOct, 0, binsPerOctave);

            // Do the top octave before resampling to allow for fast resampling
            Triple<Complex[][], Integer, double[]> fftFilterResponse = vqtFilterFFT(
                    sr, freqsOct, windowFunction, isCQT, gamma, alpha
            );
            Complex[][] fftBasis = fftFilterResponse.value0();
            int numFFT = fftFilterResponse.value1();

            // Compute the VQT filter response and append it to the list
            vqtResponses.add(vqtResponse(y, numFFT, hopLength, fftBasis));

            // Update values
            startingOctave = 1;
            filter = Filter.KAISER_FAST;

            // Update task progress
            if (task != null) task.updateProgress(1, numOctaves);
        }

        // Iterate down the octaves
        double[] myY = y;
        double mySR = sr;
        int myHopLength = hopLength;

        for (int octave = startingOctave; octave < numOctaves; octave++) {  // Starts from the HIGHEST frequencies
            // Get the frequencies of the current octave
            double[] freqsOct = new double[binsPerOctave];
            System.arraycopy(freqs, numBins - binsPerOctave * (octave + 1), freqsOct, 0, binsPerOctave);

            // Get the FFT basis and the `numFFT` for this octave
            Triple<Complex[][], Integer, double[]> fftFilterResponse = vqtFilterFFT(
                    mySR, freqsOct, windowFunction, isCQT, gamma, alpha
            );
            Complex[][] fftBasis = fftFilterResponse.value0();
            int numFFT = fftFilterResponse.value1();

            // Re-scale the filters to compensate for downsampling
            for (int i = 0; i < fftBasis.length; i++) {
                for (int j = 0; j < fftBasis[0].length; j++) {
                    fftBasis[i][j] = fftBasis[i][j].scale(Math.sqrt(sr / mySR));
                }
            }

            // Compute the VQT filter response and append to the list
            Complex[][] response = vqtResponse(myY, numFFT, myHopLength, fftBasis);
            vqtResponses.add(response);

            // Update variables
            if (myHopLength % 2 == 0) {
                myHopLength /= 2;
                mySR /= 2.;
                myY = Audio.resample(myY, 2, 1, filter, true);
            }

            // Update task progress
            if (task != null) task.updateProgress(octave + 1, numOctaves);
        }

        // Trim and stack the VQT responses
        Complex[][] V = trimAndStack(vqtResponses, numBins);

        // Get the maximum column size
        int maxCol = V[0].length;

        // Recompute lengths here because early downsampling may have changed our sampling rate
        waveletLengthsResponse = Wavelet.computeWaveletLengths(
                freqs, sr, windowFunction, 1., isCQT, gamma, alpha
        );
        double[] lengths = waveletLengthsResponse.value0();

        // Scale `V` back to normal
        for (int i = 0; i < numBins; i++) {
            double scaleFactor = Math.sqrt(lengths[i]);

            for (int j = 0; j < maxCol; j++) {
                V[i][j] = V[i][j].divides(scaleFactor);
            }
        }

        // Return VQT matrix
        log(
                Level.FINE, "VQT Matrix generated; has shape (" + V.length + ", " + V[0].length + ")", VQT.class.getName()
        );
        return V;
    }

    // Private Methods

    /**
     * Gets the number of octaves required to store <code>numBins</code>.
     *
     * @param numBins       Number of Variable-Q bins.
     * @param binsPerOctave Number of bins per octave.
     * @return Number of octaves required to store <code>numBins</code>.
     */
    private static int getNumOctaves(int numBins, int binsPerOctave) {
        return (int) Math.ceil((double) numBins / binsPerOctave);
    }

    /**
     * Perform early downsampling on the audio samples.
     *
     * @param y            Audio time series.
     * @param sr           Sample rate of the audio.
     * @param hopLength    Number of samples between successive VQT columns.
     * @param filter       Filter window.
     * @param numOctaves   Number of octaves to consider
     * @param nyquist      Nyquist frequency value.
     * @param filterCutoff Highest frequency value before cutoff begins.
     * @return Three values. First is the downsampled audio time series. Second is the new sample
     * rate. Third is the new hop length.
     * @throws ValueException If The input signal length is too short for a
     *                        <code>numOctaves</code>-octave VQT.
     */

    private static Triple<double[], Double, Integer> earlyDownsample(
            double[] y, double sr, int hopLength, Filter filter, int numOctaves, double nyquist, double filterCutoff
    ) {
        // Compute the number of early downsampling operations
        int downsampleCount1 = Math.max(0, (int) Math.ceil(MathUtils.log2(BW_FASTEST * nyquist / filterCutoff)) - 2);
        int downsampleCount2 = Math.max(0, MathUtils.numTwoFactors(hopLength) - numOctaves + 1);

        int downsampleCount = Math.min(downsampleCount1, downsampleCount2);

        // Actually perform the downsampling
        double[] yNew;

        if (filter == Filter.KAISER_FAST) {
            if (downsampleCount > 0) {
                // Compute how much to downsample by
                int downsampleFactor = (int) Math.pow(2, downsampleCount);

                // Check if the signal can actually be downsampled
                if (y.length < downsampleFactor) {
                    throw new ValueException(
                            "Input signal length of " + y.length + " is too short for " + numOctaves + "-octave VQT"
                    );
                }

                // Downsample hop length and the sample rate
                hopLength /= downsampleFactor;
                double newSr = sr / downsampleFactor;

                // Downsample audio sample
                yNew = Audio.resample(y, sr, newSr, filter, true);
                sr = newSr;
            } else {
                yNew = y;
            }
        } else {
            // Not possible for `downsampleCount > 0` due to how `downsampleCount1` is calculated
            yNew = y;
        }

        // Return needed values
        return new Triple<>(yNew, sr, hopLength);
    }

    /**
     * Generate the frequency domain variable-Q filter basis.
     *
     * @param sr             Sample rate.
     * @param freqs          Centre frequencies of the frequency bins.
     * @param windowFunction Window function to use.
     * @param isCQT          Whether this is a CQT or not.
     * @param gamma          Gamma value.
     * @param alpha          Alpha value.
     * @return Triplet of values. First value is a 2D array of complex coefficients, representing
     * the FFT basis. Second value is an integer, representing the number of FFT frequency bins.
     * Third value is a double array, representing the filters' lengths.
     */
    private static Triple<Complex[][], Integer, double[]> vqtFilterFFT(
            double sr, double[] freqs, WindowFunction windowFunction, boolean isCQT, double gamma, double alpha
    ) {
        // Get the frequency and lengths of the wavelet basis
        Pair<Complex[][], double[]> waveletBasisResponse = Wavelet.computeWaveletBasis(
                freqs, sr, windowFunction, 1, true, 1, isCQT, gamma, alpha
        );
        Complex[][] basis = waveletBasisResponse.value0();
        double[] lengths = waveletBasisResponse.value1();

        // Number of FFT bins is the second element of the shape of the basis
        int fftWindowLength = basis.length;
        int numFFT = basis[0].length;

        // Re-normalize bases with respect to the FFT window length
        for (int i = 0; i < fftWindowLength; i++) {
            double normalisationFactor = lengths[i] / numFFT;

            for (int j = 0; j < numFFT; j++) {
                basis[i][j] = basis[i][j].scale(normalisationFactor);
            }
        }

        // FFT and retain only the non-negative frequencies
        Complex[][] fftBasis = new Complex[fftWindowLength][numFFT];
        for (int i = 0; i < fftWindowLength; i++) {
            fftBasis[i] = FFT.rfft(basis[i]);
        }

        // Return required data
        return new Triple<>(fftBasis, numFFT, lengths);
    }

    /**
     * Compute the VQT filter response with a target STFT hop.
     *
     * @param y         Array of samples.
     * @param numFFT    Number of bins to use in the FFT.
     * @param hopLength Step size between each hop.
     * @param fftBasis  Basis to use for the FFT.
     * @return VQT filter response.
     */
    private static Complex[][] vqtResponse(double[] y, int numFFT, int hopLength, Complex[][] fftBasis) {
        // Get the STFT matrix
        Complex[][] D = STFT.stft(y, numFFT, hopLength, WindowFunction.ONES_WINDOW);

        // Matrix multiply `fftBasis` with `D` and return the result
        return ArrayUtils.matmul(fftBasis, D);
    }

    /**
     * Helper function to trim and stack a collection of VQT responses.
     *
     * @param vqtResponses Collection of VQT responses.
     * @param numBins      Number of bins for the VQT.
     * @return Trimmed and stacked array of VQT responses.
     */
    private static Complex[][] trimAndStack(List<Complex[][]> vqtResponses, int numBins) {
        // Get the maximum permitted number columns
        // (We take the minimum to avoid weird index errors later when processing magnitudes)
        int maxPermittedNumCol = Integer.MAX_VALUE;

        for (Complex[][] vqtResponse : vqtResponses) {
            maxPermittedNumCol = Math.min(vqtResponse[0].length, maxPermittedNumCol);
        }

        // Generate output VQT matrix
        Complex[][] vqtOut = new Complex[numBins][maxPermittedNumCol];

        // Copy per-octave data into output array
        int end = numBins;
        for (Complex[][] response : vqtResponses) {  // Note: FIRST element represents the HIGHEST FREQUENCY bin
            // By default, take the whole octave
            int numOctaves = response.length;

            // It is guaranteed that `end >= numOctaves` since `end = numBins` is a multiple of `numOctaves` and the
            // number of VQT responses is `numOctaves / numBins`, so we just copy until `numOctaves`
            for (int i = 0; i < numOctaves; i++) {
                System.arraycopy(
                        response[i],
                        0,
                        vqtOut[end - numOctaves + i],
                        0,
                        maxPermittedNumCol
                );
            }

            end -= numOctaves;
        }

        // Return the stacked VQT matrix
        return vqtOut;
    }
}
